src/category_theory/pseudoabelian/basic.lean

```diff
@@ -0,0 +1,314 @@
+/-
+Copyright (c) 2022 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import category_theory.preadditive.additive_functor
+import category_theory.equivalence
+
+/-!
+# Pseudoabelian categories
+
+In this file, we define the Karoubi envelope `karoubi C` of a preadditive
+category `C` and the notion of pseudoabelian category (also known as Karoubian categories).
+
+## Main constructions
+
+- `is_pseudoabelian C` expresses that `C` is pseudoabelian, i.e. all idempotents endomorphisms
+in `C` have a kernel.
+- `karoubi C` is the pseudoabelian or Karoubi envelope of a preadditive category `C`.
+- `karoubi_is_pseudoabelian C` says that `karoubi C` is pseudoabelian.
+- `to_karoubi C : C ⥤ karoubi C` is a fully faithful functor, which is an equivalence
+(`to_karoubi_is_equivalence`) when `C` is pseudoabelian.
+
+## References
+* [Stacks: Karoubian categories] https://stacks.math.columbia.edu/tag/09SF
+
+-/
+
+noncomputable theory
+
+open category_theory.category
+open category_theory.preadditive
+open category_theory.limits
+open_locale big_operators
+
+namespace category_theory
+
+variables (C : Type*) [category C] [preadditive C]
+
+namespace pseudoabelian
+
+/-- When an object `X` decomposes as `X ≅ P ⨿ Q`, one may consider `P` as a direct factor of `X`
+and up to unique isomorphism, it is determined by the obvious idempotent `X ⟶ P ⟶ X` which is the
+projector on `P` with kernel `Q`. More generally, one may define a formal director of an object
+`X : C` : it consists of an idempotent `p : X ⟶ X` which is thought as the "formal image" of `p`.
+The type `karoubi C` shall be the objects of the karoubi enveloppe of `C`. -/
+@[nolint has_inhabited_instance]
+structure karoubi := (X : C) (p : X ⟶ X) (idempotence : p ≫ p = p)
+
+end pseudoabelian
+
+/-- A preadditive category is pseudoabelian iff all idempotent endomorphisms have a kernel. -/
+class is_pseudoabelian : Prop :=
+(idempotents_have_kernels : Π (P : pseudoabelian.karoubi C), has_kernel P.p)
+
+namespace pseudoabelian
+
+namespace karoubi
+
+variables {C}
+
+@[ext]
+lemma ext {P Q : karoubi C} (h_X : P.X = Q.X)
+  (h_p : P.p ≫ eq_to_hom h_X = eq_to_hom h_X ≫ Q.p) : P = Q :=
```

Comment:
I'm a bit wary of proving equality of objects in a category. Do you need this a lot?

Reply:
I really need this. In the Dold-Kan equivalence, I need some equalities of functors (like a composition of 4 functors is equal to some other composition); using isomorphisms instead would be highly unpractical.